File: iOS-ObjC/watf-bank/PageController/Pin.m
#import <LocalAuthentication/LocalAuthentication.h>
#import "Util.h"

@interface Pin:UIViewController <UITextFieldDelegate>
@property (weak, nonatomic) IBOutlet UITextField *pin;
@property (weak, nonatomic) IBOutlet UILabel *label;
@end

@implementation Pin

bool fingerprint = true;

- (void)viewDidLoad {
    [super viewDidLoad];
    _pin.delegate = self;
    if([Util getKeychain:@"pin"]!=nil)
    {
        _label.text = @"Enter Your PIN";
    }
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    NSCharacterSet *numbersOnly = [NSCharacterSet characterSetWithCharactersInString:@"0123456789"];
    NSCharacterSet *characterSetFromTextField = [NSCharacterSet characterSetWithCharactersInString:string];
    BOOL stringIsValid = [numbersOnly isSupersetOfSet:characterSetFromTextField];
    BOOL lengthIsValid = ([textField.text length]+[string length]>4);
    return stringIsValid && !lengthIsValid;
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [_pin resignFirstResponder];
}


- (void)viewWillAppear:(BOOL)animated
{
    if([Util getKeychain:@"pin"]==nil)
    {
        return;
    }
    if(fingerprint)
    {
        LAContext *context = [[LAContext alloc] init];
        NSError *err = nil;
        NSString *reason = @"Authentication";
        fingerprint = false;
        if ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&err]) {
            [context evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
                      localizedReason:reason
                                reply:^(BOOL success, NSError *error) {
                                    dispatch_async(dispatch_get_main_queue(), ^{
                                        if (success)
                                        {
                                            fingerprint = true;
                                            [Util changeView:self :@"Home"];
                                        }
                                        else if(error.code==-1)
                                        {
                                            [Util alert:self :@"Too many failed attemps"];
                                        }
                                        else if(error.code!=-2&&error.code!=-3)
                                        {
                                            [Util alert:self :@"An Error Occured"];
                                        }
                                    });
                                }];
        } else {
            [Util alert:self :@"Check Your Touch ID"];
        }
    }
}

- (IBAction)push:(id)sender {
    if([Util getKeychain:@"pin"]!=nil)
    {
        if([self checkPin])
        {
            fingerprint = true;
            [self nextPage];
        }
        else
        {
            [Util alert:self :@"Invalid PIN"];
        }
    }
    else
    {
        if([_pin.text length]!=4)
        {
            [Util alert:self :@"Please Insert 4-Digit Pin"];
        }
        else if([Util saveKeychain:@"pin" :[Util md5:_pin.text]])
        {
            fingerprint = true;
            [self nextPage];
        }
        else
        {
            [Util alert:self :@"An Error Occured"];
        }
    }
}

-(BOOL)checkPin
{
    return [[Util getKeychain:@"pin"] isEqualToString:[Util md5:_pin.text]];
}

-(void)nextPage
{
    [Util changeView:self :@"Home"];
}

@end

File: iOS-Swift/watf-bank/PageController/Pin.swift
import UIKit
import LocalAuthentication
import CommonCrypto

class Pin : UIViewController, UITextFieldDelegate {
    
    @IBOutlet weak var pinText: UITextField!
    @IBOutlet weak var label: UILabel!
    
    override func viewDidLoad() {
        pinText.delegate = self
        if KeyChain.load("pin") != nil {
            label.text = "Enter PIN"
        }
    }
    
    @IBAction func push(_ sender: Any) {
        if let pin = KeyChain.load("pin") {
            if pin == md5(pinText.text!) {
                Util.changeView(self, "Home")
            }
            else {
                Util.alert(self, "Invalid PIN")
            }
        }
        else {
            if pinText.text!.count == 4 {
                if !KeyChain.save("pin", md5(pinText.text!)) {
                    Util.alert(self, "An Error Occured")
                }
                else {
                    Util.changeView(self, "Home")
                }
            }
            else {
                Util.alert(self, "Please Insert 4-digit PIN")
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        if KeyChain.load("pin") == nil {
            let context = LAContext()
            var error: NSError?
            if context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) {
                context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: "Authentication") {
                    [unowned self] success, authenticationError in
                    DispatchQueue.main.async {
                        if success {
                            Util.changeView(self, "Home")
                        }
                        else {
                            Util.alert(self, "Too Many Errors")
                        }
                    }
                }
            }
            else {
                Util.alert(self, "Check Your Touch ID")
            }
        }
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.resignFirstResponder()
    }
    
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        let allowChar = CharacterSet(charactersIn: "0123456789")
        return  allowChar.isSuperset(of: CharacterSet(charactersIn: string)) && (textField.text!.count+string.count)<=4
    }
    
    func md5(_ text: String) -> String {
        let messageData = text.data(using:.utf8)!
        var digestData = Data(count: Int(CC_MD5_DIGEST_LENGTH))
        
        _ = digestData.withUnsafeMutableBytes {digestBytes in
            messageData.withUnsafeBytes {messageBytes in
                CC_MD5(messageBytes, CC_LONG(messageData.count), digestBytes)
            }
        }
        return digestData.base64EncodedString()
    }
}
